File: app/src/main/java/com/cadnative/firevisioniptv/ChannelManager.java
package com.cadnative.firevisioniptv;

import android.content.ComponentName;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.media.tv.TvContract;
import android.net.Uri;
import android.util.Log;

import androidx.tvprovider.media.tv.TvContractCompat;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.realm.Realm;

/**
 * Manages channel synchronization with Fire TV's TIF database
 * Handles inserting, updating, and removing channels
 */
public class ChannelManager {

    private static final String TAG = "ChannelManager";
    private Context mContext;
    private String mInputId;

    public ChannelManager(Context context) {
        mContext = context;
        mInputId = TvContract.buildInputId(
            new ComponentName(context, FireVisionTvInputService.class)
        );
    }

    /**
     * Sync all entitled channels to Fire TV TIF database
     */
    public void syncChannelsToTif() {
        Log.d(TAG, "Starting channel sync to TIF database");

        // Get current channels from your database (Realm)
        List<Channel> appChannels = getChannelsFromRealm();

        // Get existing TIF channels
        Map<String, Long> existingTifChannels = getExistingTifChannels();

        // Insert or update channels
        for (Channel channel : appChannels) {
            String channelId = channel.getChannelId();

            if (existingTifChannels.containsKey(channelId)) {
                // Update existing channel
                updateChannel(existingTifChannels.get(channelId), channel);
            } else {
                // Insert new channel
                insertChannel(channel);
            }
        }

        Log.d(TAG, "Channel sync completed. Synced " + appChannels.size() + " channels");
    }

    /**
     * Get channels from Realm database
     */
    private List<Channel> getChannelsFromRealm() {
        List<Channel> channels = new ArrayList<>();

        try {
            Realm realm = Realm.getDefaultInstance();
            List<Channel> realmChannels = realm.where(Channel.class).findAll();

            // Copy to regular list
            for (Channel channel : realmChannels) {
                Channel copy = new Channel();
                copy.setChannelId(channel.getChannelId());
                copy.setChannelName(channel.getChannelName());
                copy.setChannelUrl(channel.getChannelUrl());
                copy.setChannelImg(channel.getChannelImg());
                copy.setChannelGroup(channel.getChannelGroup());
                channels.add(copy);
            }

            realm.close();
        } catch (Exception e) {
            Log.e(TAG, "Error getting channels from Realm", e);
        }

        return channels;
    }

    /**
     * Get existing channels from TIF database
     */
    private Map<String, Long> getExistingTifChannels() {
        Map<String, Long> existingChannels = new HashMap<>();
        ContentResolver resolver = mContext.getContentResolver();

        Uri channelsUri = TvContract.buildChannelsUriForInput(mInputId);
        String[] projection = {
            TvContract.Channels._ID,
            TvContract.Channels.COLUMN_INTERNAL_PROVIDER_DATA
        };

        try (Cursor cursor = resolver.query(channelsUri, projection, null, null, null)) {
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    long id = cursor.getLong(0);
                    String internalData = cursor.getString(1);

                    // Extract channel ID from internal data
                    try {
                        JSONObject json = new JSONObject(internalData);
                        String channelId = json.optString("channelId");
                        if (!channelId.isEmpty()) {
                            existingChannels.put(channelId, id);
                        }
                    } catch (JSONException e) {
                        Log.e(TAG, "Error parsing internal data", e);
                    }
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Error querying TIF channels", e);
        }

        return existingChannels;
    }

    /**
     * Insert a new channel into TIF database
     */
    private Uri insertChannel(Channel channel) {
        ContentResolver resolver = mContext.getContentResolver();

        try {
            // Build internal provider data with channel info and deep link
            JSONObject internalData = new JSONObject();
            internalData.put("channelId", channel.getChannelId());
            internalData.put("channelUrl", channel.getChannelUrl());

            // Create deep link for playback
            Intent playIntent = new Intent(Intent.ACTION_VIEW);
            playIntent.setComponent(new ComponentName(mContext, PlaybackActivity.class));
            playIntent.putExtra("channel", channel);
            String deepLinkUri = playIntent.toUri(Intent.URI_INTENT_SCHEME);
            internalData.put("playbackDeepLinkUri", deepLinkUri);

            // Build channel content values
            ContentValues values = new ContentValues();
            values.put(TvContract.Channels.COLUMN_INPUT_ID, mInputId);
            values.put(TvContract.Channels.COLUMN_TYPE, TvContract.Channels.TYPE_OTHER);
            values.put(TvContract.Channels.COLUMN_DISPLAY_NAME, truncateName(channel.getChannelName()));
            values.put(TvContract.Channels.COLUMN_DISPLAY_NUMBER, channel.getChannelId());
            values.put(TvContract.Channels.COLUMN_INTERNAL_PROVIDER_DATA, internalData.toString());

            // Add channel logo if available
            if (channel.getChannelImg() != null && !channel.getChannelImg().isEmpty()) {
                values.put(TvContractCompat.Channels.COLUMN_APP_LINK_ICON_URI, channel.getChannelImg());
            }

            Uri channelUri = resolver.insert(TvContract.Channels.CONTENT_URI, values);
            Log.d(TAG, "Inserted channel: " + channel.getChannelName() + " -> " + channelUri);

            return channelUri;

        } catch (Exception e) {
            Log.e(TAG, "Error inserting channel: " + channel.getChannelName(), e);
            return null;
        }
    }

    /**
     * Update an existing channel in TIF database
     */
    private void updateChannel(long tifChannelId, Channel channel) {
        ContentResolver resolver = mContext.getContentResolver();

        try {
            Uri channelUri = TvContract.buildChannelUri(tifChannelId);

            ContentValues values = new ContentValues();
            values.put(TvContract.Channels.COLUMN_DISPLAY_NAME, truncateName(channel.getChannelName()));

            if (channel.getChannelImg() != null && !channel.getChannelImg().isEmpty()) {
                values.put(TvContractCompat.Channels.COLUMN_APP_LINK_ICON_URI, channel.getChannelImg());
            }

            resolver.update(channelUri, values, null, null);
            Log.d(TAG, "Updated channel: " + channel.getChannelName());

        } catch (Exception e) {
            Log.e(TAG, "Error updating channel: " + channel.getChannelName(), e);
        }
    }

    /**
     * Remove all channels from TIF database
     */
    public void removeAllChannels() {
        ContentResolver resolver = mContext.getContentResolver();
        Uri channelsUri = TvContract.buildChannelsUriForInput(mInputId);

        int deleted = resolver.delete(channelsUri, null, null);
        Log.d(TAG, "Removed " + deleted + " channels from TIF");
    }

    /**
     * Truncate channel name to 25 characters (Fire TV requirement)
     */
    private String truncateName(String name) {
        if (name == null) return "Unknown";
        if (name.length() <= 25) return name;
        return name.substring(0, 22) + "...";
    }

    /**
     * Trigger channel initialization broadcast
     */
    public void triggerChannelUpdate() {
        Intent intent = new Intent(TvContract.ACTION_INITIALIZE_PROGRAMS);
        intent.setComponent(new ComponentName(mContext, ChannelUpdateReceiver.class));
        mContext.sendBroadcast(intent);
        Log.d(TAG, "Triggered channel update broadcast");
    }
}

File: app/src/main/java/com/cadnative/firevisioniptv/ChannelManager.kt
package com.cadnative.firevisioniptv

import android.content.ComponentName
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.media.tv.TvContract
import android.net.Uri
import android.util.Log
import androidx.tvprovider.media.tv.TvContractCompat
import com.cadnative.firevisioniptv.data.source.local.dao.ChannelDao
import com.cadnative.firevisioniptv.data.source.local.entity.ChannelEntity
import com.cadnative.firevisioniptv.domain.repository.EpgRepository
import dagger.hilt.EntryPoint
import dagger.hilt.InstallIn
import dagger.hilt.android.EntryPointAccessors
import dagger.hilt.components.SingletonComponent
import org.json.JSONObject

/**
 * Manages channel and EPG synchronization with Fire TV's TIF database.
 * Uses Hilt EntryPoint to access DAO/repositories from non-injected components.
 */
class ChannelManager private constructor(
    private val context: Context,
    private val channelDao: ChannelDao,
    private val epgRepository: EpgRepository
) {

    @EntryPoint
    @InstallIn(SingletonComponent::class)
    interface ChannelManagerEntryPoint {
        fun channelDao(): ChannelDao
        fun epgRepository(): EpgRepository
    }

    private val inputId: String = TvContract.buildInputId(
        ComponentName(context, FireVisionTvInputService::class.java)
    )

    /**
     * Sync all active channels to Fire TV TIF database, then sync EPG.
     */
    suspend fun syncChannelsToTif() {
        Log.d(TAG, "Starting channel sync to TIF database")

        val appChannels = channelDao.getAllActiveChannels()
        val existingTifChannels = getExistingTifChannels()
        val appChannelIds = appChannels.map { it.id }.toSet()

        // Remove TIF channels that no longer exist in app
        for ((channelId, tifId) in existingTifChannels) {
            if (channelId !in appChannelIds) {
                try {
                    context.contentResolver.delete(TvContract.buildChannelUri(tifId), null, null)
                    Log.d(TAG, "Removed stale TIF channel: $channelId")
                } catch (e: Exception) {
                    Log.e(TAG, "Error removing stale channel: $channelId", e)
                }
            }
        }

        // Insert or update channels
        for (channel in appChannels) {
            val existingId = existingTifChannels[channel.id]
            if (existingId != null) {
                updateChannel(existingId, channel)
            } else {
                insertChannel(channel)
            }
        }

        Log.d(TAG, "Channel sync completed. Synced ${appChannels.size} channels")

        syncEpgToTif()
    }

    /**
     * Sync EPG program data to TvContract.Programs for channels with a tvgId.
     */
    private suspend fun syncEpgToTif() {
        Log.d(TAG, "Starting EPG sync to TIF")

        try {
            epgRepository.ensureLoaded()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load EPG data", e)
            return
        }

        val tifChannels = getExistingTifChannels()
        val appChannels = channelDao.getAllActiveChannels()
        var programCount = 0

        for (channel in appChannels) {
            val tvgId = channel.tvgId ?: continue
            val tifChannelId = tifChannels[channel.id] ?: continue

            try {
                val programsUri = TvContract.buildProgramsUriForChannel(tifChannelId)
                context.contentResolver.delete(programsUri, null, null)

                val (now, next) = epgRepository.getNowNext(tvgId)

                listOfNotNull(now, next).forEach { program ->
                    val values = ContentValues().apply {
                        put(TvContract.Programs.COLUMN_CHANNEL_ID, tifChannelId)
                        put(TvContract.Programs.COLUMN_TITLE, program.title)
                        program.description?.let {
                            put(TvContract.Programs.COLUMN_SHORT_DESCRIPTION, it)
                        }
                        put(TvContract.Programs.COLUMN_START_TIME_UTC_MILLIS, program.startTime.toEpochMilli())
                        put(TvContract.Programs.COLUMN_END_TIME_UTC_MILLIS, program.endTime.toEpochMilli())
                        program.icon?.let {
                            put(TvContract.Programs.COLUMN_POSTER_ART_URI, it)
                        }
                    }
                    context.contentResolver.insert(TvContract.Programs.CONTENT_URI, values)
                    programCount++
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error syncing EPG for channel: ${channel.name}", e)
            }
        }

        Log.d(TAG, "EPG sync completed. Synced $programCount programs")
    }

    private fun getExistingTifChannels(): Map<String, Long> {
        val existingChannels = mutableMapOf<String, Long>()
        val resolver = context.contentResolver

        val channelsUri = TvContract.buildChannelsUriForInput(inputId)
        val projection = arrayOf(
            TvContract.Channels._ID,
            TvContract.Channels.COLUMN_INTERNAL_PROVIDER_DATA
        )

        try {
            resolver.query(channelsUri, projection, null, null, null)?.use { cursor ->
                while (cursor.moveToNext()) {
                    val id = cursor.getLong(0)
                    val internalData = cursor.getString(1)

                    try {
                        val json = JSONObject(internalData ?: "{}")
                        val channelId = json.optString("channelId")
                        if (channelId.isNotEmpty()) {
                            existingChannels[channelId] = id
                        }
                    } catch (e: Exception) {
                        Log.e(TAG, "Error parsing internal data", e)
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error querying TIF channels", e)
        }

        return existingChannels
    }

    private fun insertChannel(channel: ChannelEntity): Uri? {
        val resolver = context.contentResolver

        return try {
            val internalData = JSONObject().apply {
                put("channelId", channel.id)
                put("channelUrl", channel.streamUrl)
                put("playbackDeepLinkUri", "firevisioniptv://play/channel/${channel.id}")
            }

            val values = ContentValues().apply {
                put(TvContract.Channels.COLUMN_INPUT_ID, inputId)
                put(TvContract.Channels.COLUMN_TYPE, TvContract.Channels.TYPE_OTHER)
                put(TvContract.Channels.COLUMN_DISPLAY_NAME, truncateName(channel.name))
                put(TvContract.Channels.COLUMN_DISPLAY_NUMBER, channel.id)
                put(TvContract.Channels.COLUMN_INTERNAL_PROVIDER_DATA, internalData.toString())
                channel.logoUrl?.takeIf { it.isNotEmpty() }?.let {
                    put(TvContractCompat.Channels.COLUMN_APP_LINK_ICON_URI, it)
                }
            }

            val channelUri = resolver.insert(TvContract.Channels.CONTENT_URI, values)
            Log.d(TAG, "Inserted channel: ${channel.name} -> $channelUri")
            channelUri
        } catch (e: Exception) {
            Log.e(TAG, "Error inserting channel: ${channel.name}", e)
            null
        }
    }

    private fun updateChannel(tifChannelId: Long, channel: ChannelEntity) {
        val resolver = context.contentResolver

        try {
            val channelUri = TvContract.buildChannelUri(tifChannelId)

            val internalData = JSONObject().apply {
                put("channelId", channel.id)
                put("channelUrl", channel.streamUrl)
                put("playbackDeepLinkUri", "firevisioniptv://play/channel/${channel.id}")
            }

            val values = ContentValues().apply {
                put(TvContract.Channels.COLUMN_DISPLAY_NAME, truncateName(channel.name))
                put(TvContract.Channels.COLUMN_INTERNAL_PROVIDER_DATA, internalData.toString())
                channel.logoUrl?.takeIf { it.isNotEmpty() }?.let {
                    put(TvContractCompat.Channels.COLUMN_APP_LINK_ICON_URI, it)
                }
            }

            resolver.update(channelUri, values, null, null)
            Log.d(TAG, "Updated channel: ${channel.name}")
        } catch (e: Exception) {
            Log.e(TAG, "Error updating channel: ${channel.name}", e)
        }
    }

    fun removeAllChannels() {
        val resolver = context.contentResolver
        val channelsUri = TvContract.buildChannelsUriForInput(inputId)
        val deleted = resolver.delete(channelsUri, null, null)
        Log.d(TAG, "Removed $deleted channels from TIF")
    }

    private fun truncateName(name: String?): String {
        if (name == null) return "Unknown"
        if (name.length <= 25) return name
        return "${name.substring(0, 22)}..."
    }

    fun triggerChannelUpdate() {
        val intent = Intent(TvContract.ACTION_INITIALIZE_PROGRAMS).apply {
            component = ComponentName(context, ChannelUpdateReceiver::class.java)
        }
        context.sendBroadcast(intent)
        Log.d(TAG, "Triggered channel update broadcast")
    }

    companion object {
        private const val TAG = "ChannelManager"

        fun create(context: Context): ChannelManager {
            val appContext = context.applicationContext
            val entryPoint = EntryPointAccessors.fromApplication(
                appContext, ChannelManagerEntryPoint::class.java
            )
            return ChannelManager(
                context = appContext,
                channelDao = entryPoint.channelDao(),
                epgRepository = entryPoint.epgRepository()
            )
        }
    }
}
